`opennebula_service`: Plugin did not respond

### Description

Hello,

I am trying to deploy a service from Terraform but I find this error. Could you help me?

```
opennebula_service.ubuntu_service: Creating...
╷
│ Error: Plugin did not respond
│ 
│   with opennebula_service.ubuntu_service,
│   on one_create_service.tf line 18, in resource "opennebula_service" "ubuntu_service":
│   18: resource "opennebula_service" "ubuntu_service" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more
│ details.
╵
```

### Terraform and Provider version

```
Terraform v1.4.2
on darwin_arm64
+ provider registry.terraform.io/opennebula/opennebula v1.2.0
```

### Affected resources and data sources

`opennebula_service`

### Terraform configuration

```hcl
resource "opennebula_service" "ubuntu_service" {
  name           = "service-template-test"
  template_id    = 2
}
```

### Expected behavior

Service to be deployed

### Actual behavior

The terraform-provider-opennebula_v1.2.0 plugin crashed!

### Steps to Reproduce

terraform init
terraform apply

### Debug output

_No response_

### Panic output

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x18 pc=0x100737540]

goroutine 51 [running]:
github.com/OpenNebula/one/src/oca/go/src/goca.(*STemplateController).Instantiate(0x14000636ba8, {0x0, 0x0})
        github.com/OpenNebula/one/src/oca/go/src/goca@v0.0.0-20230301133003-197f04efa071/service_template.go:159 +0x260
github.com/OpenNebula/terraform-provider-opennebula/opennebula.resourceOpennebulaServiceCreate({0x100c867a8, 0x140004a5320}, 0x14000564e00, {0x100b3ba60?, 0x14000516980})
        github.com/OpenNebula/terraform-provider-opennebula/opennebula/resource_opennebula_service.go:164 +0xe4
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).create(0x14000476380, {0x100c867e0, 0x1400055ecf0}, 0xd?, {0x100b3ba60, 0x14000516980})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/resource.go:707 +0xec
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0x14000476380, {0x100c867e0, 0x1400055ecf0}, 0x14000611ad0, 0x14000564b80, {0x100b3ba60, 0x14000516980})
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/resource.go:837 +0x874
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0x14000205758, {0x100c867e0?, 0x1400055ebd0?}, 0x1400007d090)
        github.com/hashicorp/terraform-plugin-sdk/v2@v2.24.1/helper/schema/grpc_provider.go:1021 +0xb94
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0x1400035d040, {0x100c867e0?, 0x1400055e3c0?}, 0x14000368850)
        github.com/hashicorp/terraform-plugin-go@v0.14.1/tfprotov5/tf5server/server.go:818 +0x3c0
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler({0x100c4c1c0?, 0x1400035d040}, {0x100c867e0, 0x1400055e3c0}, 0x140003687e0, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.14.1/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:385 +0x174
google.golang.org/grpc.(*Server).processUnaryRPC(0x140003641e0, {0x100c897e0, 0x14000489380}, 0x1400018e000, 0x140004901e0, 0x10115a6a0, 0x0)
        google.golang.org/grpc@v1.50.1/server.go:1340 +0xb90
google.golang.org/grpc.(*Server).handleStream(0x140003641e0, {0x100c897e0, 0x14000489380}, 0x1400018e000, 0x0)
        google.golang.org/grpc@v1.50.1/server.go:1713 +0x840
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        google.golang.org/grpc@v1.50.1/server.go:965 +0x88
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.50.1/server.go:963 +0x298
```

### Important factoids

_No response_

### References

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -17,6 +17,8 @@ BUG FIXES:
   * `mtu`, `guest_mtu`, `description`, `gateway`, `network_mask`, `network_address`, `search_domain`, `dns` are not marked as `computed` and are not read for a reservation virtual network
   * `gateway`, `network_mask`, `network_address`, `search_domain`, `dns` are now read by the provider
   * When not used more emtpy values are set for attributes
+* resources/opennebula_service: check if flow client is nil (#445)
+* resources/opennebula_service_template: check if flow client is nil (#445)
 
 # 1.2.0 (March 23th, 2023)
 

--- a/opennebula/provider.go
+++ b/opennebula/provider.go
@@ -118,6 +118,10 @@ type Configuration struct {
 	newDefaultTags map[string]interface{}
 }
 
+func (c *Configuration) isFlowConfigured() bool {
+	return c.Controller.ClientFlow != nil
+}
+
 func providerConfigure(ctx context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
 
 	var diags diag.Diagnostics

--- a/opennebula/resource_opennebula_service.go
+++ b/opennebula/resource_opennebula_service.go
@@ -148,9 +148,19 @@ func resourceOpennebulaServiceCreate(ctx context.Context, d *schema.ResourceData
 	config := meta.(*Configuration)
 	controller := config.Controller
 
+	var diags diag.Diagnostics
+
+	if !config.isFlowConfigured() {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Flow client isn't configured",
+			Detail:   fmt.Sprintf("Check flow_endpoint in the provider configuration"),
+		})
+		return diags
+	}
+
 	var err error
 	var serviceID int
-	var diags diag.Diagnostics
 
 	// if template id is set, instantiate a Service from this template
 	tc := controller.STemplate(d.Get("template_id").(int))
@@ -240,7 +250,18 @@ func resourceOpennebulaServiceCreate(ctx context.Context, d *schema.ResourceData
 
 func resourceOpennebulaServiceRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 
+	config := meta.(*Configuration)
+
 	var diags diag.Diagnostics
+
+	if !config.isFlowConfigured() {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Flow client isn't configured",
+			Detail:   fmt.Sprintf("Check flow_endpoint in the provider configuration"),
+		})
+		return diags
+	}
 
 	sc, err := getServiceController(d, meta)
 	if err != nil {
@@ -312,7 +333,18 @@ func resourceOpennebulaServiceRead(ctx context.Context, d *schema.ResourceData, 
 
 func resourceOpennebulaServiceDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 
+	config := meta.(*Configuration)
+
 	var diags diag.Diagnostics
+
+	if !config.isFlowConfigured() {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Flow client isn't configured",
+			Detail:   fmt.Sprintf("Check flow_endpoint in the provider configuration"),
+		})
+		return diags
+	}
 
 	//Get Service
 	sc, err := getServiceController(d, meta)
@@ -376,10 +408,20 @@ func resourceOpennebulaServiceExists(d *schema.ResourceData, meta interface{}) (
 }
 
 func resourceOpennebulaServiceUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+
 	config := meta.(*Configuration)
 	controller := config.Controller
 
 	var diags diag.Diagnostics
+
+	if !config.isFlowConfigured() {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Flow client isn't configured",
+			Detail:   fmt.Sprintf("Check flow_endpoint in the provider configuration"),
+		})
+		return diags
+	}
 
 	//Get Service controller
 	sc, err := getServiceController(d, meta)

--- a/opennebula/resource_opennebula_service_template.go
+++ b/opennebula/resource_opennebula_service_template.go
@@ -116,6 +116,15 @@ func resourceOpennebulaServiceTemplateCreate(ctx context.Context, d *schema.Reso
 
 	var diags diag.Diagnostics
 
+	if !config.isFlowConfigured() {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Flow client isn't configured",
+			Detail:   fmt.Sprintf("Check flow_endpoint in the provider configuration"),
+		})
+		return diags
+	}
+
 	// Marshall the json
 	stemplate := &srv_tmpl.ServiceTemplate{}
 	err := json.Unmarshal([]byte(d.Get("template").(string)), stemplate)
@@ -195,8 +204,18 @@ func resourceOpennebulaServiceTemplateCreate(ctx context.Context, d *schema.Reso
 }
 
 func resourceOpennebulaServiceTemplateRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*Configuration)
 
 	var diags diag.Diagnostics
+
+	if !config.isFlowConfigured() {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Flow client isn't configured",
+			Detail:   fmt.Sprintf("Check flow_endpoint in the provider configuration"),
+		})
+		return diags
+	}
 
 	stc, err := getServiceTemplateController(d, meta)
 	if err != nil {
@@ -251,7 +270,18 @@ func resourceOpennebulaServiceTemplateRead(ctx context.Context, d *schema.Resour
 
 func resourceOpennebulaServiceTemplateDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 
+	config := meta.(*Configuration)
+
 	var diags diag.Diagnostics
+
+	if !config.isFlowConfigured() {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Flow client isn't configured",
+			Detail:   fmt.Sprintf("Check flow_endpoint in the provider configuration"),
+		})
+		return diags
+	}
 
 	//Get Service
 	stc, err := getServiceTemplateController(d, meta)
@@ -301,6 +331,15 @@ func resourceOpennebulaServiceTemplateUpdate(ctx context.Context, d *schema.Reso
 	controller := config.Controller
 
 	var diags diag.Diagnostics
+
+	if !config.isFlowConfigured() {
+		diags = append(diags, diag.Diagnostic{
+			Severity: diag.Error,
+			Summary:  "Flow client isn't configured",
+			Detail:   fmt.Sprintf("Check flow_endpoint in the provider configuration"),
+		})
+		return diags
+	}
 
 	//Get Service controller
 	stc, err := getServiceTemplateController(d, meta)

